### PR TITLE
Reconcile Sub Component Limitador CR

### PR DESCRIPTION
**Pull request by Boomatang**
Part of: https://github.com/Kuadrant/kuadrant-operator/issues/307
Closes: https://github.com/Kuadrant/kuadrant-operator/issues/163
RFC: https://github.com/Kuadrant/architecture/blob/main/rfcs/0006-kuadrant_sub_components_configurations.md

The changes here allow a user to configure the setting in the Limitador CR from the kuadrant CR. 

# Verification
* Deploy this branch using `make local-setup`.
* Wait for deploy to complete and all resources to be a ready state.
* Deploy the custom manifest.
  * `kubectl apply -f https://raw.githubusercontent.com/Boomatang/k8s_configs/kuadrant-operator-PR350/manifest/kuadrant_spec_limitador.yaml
`
  * The custom manifest will deploy an instance of redis along with all the resources to test the different configurations.
  * The Kuadrant CR has every [spec](https://github.com/Boomatang/k8s_configs/blob/kuadrant-operator-PR350/manifest/kuadrant_spec_limitador.yaml#L129-L158) that this PR allows configuring set.
  * The manifest is linked to a tag [kuadrant-operator-PR350](https://github.com/Boomatang/k8s_configs/tree/kuadrant-operator-PR350) to ensure state does not change.
* Wait of deployment to complete
* The spec that is being reconciled are listed in the RFC: [Limitador Spec](https://github.com/Kuadrant/architecture/blob/main/rfcs/0006-kuadrant_sub_components_configurations.md#limitador-spec)
* Editing the spec directly  in the Limitador CR should be reverted to the spec in the Kuadrant CR
* Editing the spec in the Kuadrant CR should update the spec in the Limitador CR
  * **NOTE:** fully removing a spec from the kuadrant CR will not remove it from the Limitador CR. This is do to allow upgrades. A user may have custom spec already in the Limitador CR. This change only overrides the Limitador CR spec if that spec is defined in Kuadrant CR.

### Existing Limitador Users
It is possible that a user can be using limitador as a standalone product before using kuadrant. In the case where the user named the limitador CR `limitador` and places the kuadrant CR in the same namespace, the kuadrant operator will take ownership of the limitador CR.  This will be done by setting the ownerReference in the limitador CR.

Validating this use case.
* Follow the verification steps to get the environment set up.
* Scale down the kuadrant operator
* Edit the limitador CR 
  * Set spec as follows
  ```yaml
  spec:
    replicas: 1
  ```
  * Remove the `ownerReference` block from the metadata.
* Scale back up the kuadrant operator
* Expected:
  * The `spec` switches to that defined in the kuadrant CR
  * The `ownerReference` is set to the kuadrant operator

### CODE REVIEW

File: api/v1beta1/kuadrant_types.go [hunk 1]
```diff
@@ -29,6 +31,26 @@ import (
 
 // KuadrantSpec defines the desired state of Kuadrant
 type KuadrantSpec struct {
+	// +optional
+	Limitador *LimitadorSpec `json:"limitador,omitempty"`
```
Comment (guicassolato) [inline api/v1beta1/kuadrant_types.go:35]: I suppose we don't want the entire [`limitadorv1alpha1.LimitadorSpec`](https://github.com/Kuadrant/limitador-operator/blob/e0517a252dd7a96fa117adfb443e356a6233eb6d/api/v1alpha1/limitador_types.go#L61) type here, right?

Comment (Boomatang) [reply api/v1beta1/kuadrant_types.go:35]: Correct. Some spec may be forced by the operator and not user editable. While others are classified as dev/test and should not be easily set in the kuadrant operator.

**Comment by KevFan** [comment]
@Boomatang Can you rebase with main? Running `make local-setup` is failing with:
```sh
The Istio "istiocontrolplane" is invalid: spec.namespace: Required value
make[4]: *** [sail-install] Error 1
make[3]: *** [istio-install] Error 2
make[2]: *** [local-cluster-setup] Error 2
make[1]: *** [local-env-setup] Error 2
make: *** [local-setup] Error 2
```

File: controllers/kuadrant_controller.go [hunk 2]
```diff
@@ -469,27 +471,69 @@ func (r *KuadrantReconciler) removeAnnotationFromGateways(ctx context.Context, k
...
+	if err != nil {
+		return err
+	}
+
+	if limitador.OwnerReferences != nil && !reflect.DeepEqual(tmp.OwnerReferences, limitador.OwnerReferences) {
+		return nil
 	}
 
-	err := r.SetOwnerReference(kObj, limitador)
+	err = r.SetOwnerReference(kObj, limitador)
```
[the remark this answers is not in the document]
Comment (Boomatang) [reply controllers/kuadrant_controller.go:531]: There is a bug here in my mutate function, working on the fix now, but should explain what is going on here. 

Firstly the Kuadrant operator expects the limitador CR to be called `limitador` and in the same namespace as the Kuadrant CR. The code will not reconcile any limitador resources of a different name or in different namespace. 

The `reflect.DeepEqual` on line 527 is me being extra safe. I would only expect that to true if the user has its own custom controller that is managing the limitador CR. I could see this being a thing if the user was to use limitador as standalone product before using kuadrant. 

So the bug is in the setting of the owner reference if the user manually creates the limitador CR in the same namespace as the kuadrant CR when the kuadrant operator is scaled down for example. Or the real world use case would be if the user was using the limitador as a standalone product and then starts to use kuadrant. In this case you would expect kuadrant to take ownership of the limitador CR.

[the remark this answers is not in the document]
Comment (KevFan) [reply controllers/kuadrant_controller.go:531]: Yeah, this is an edge case. I guess to me, if I were to use limitador standalone first (using the `limitador` name) in the kuadrant namespace, having kuadrant take ownership of the CR afterwards may not necessary be expected, as this limitador could be used by other services 🤔 Deleting the kuadrant CR, will then delete this limitador instance, potentially affecting other services using this limitador instance 🤔 

Anyhow, just a thought, I don't mind as long as this is expected flow, though maybe best documented somewhere that this will happen at least.

In the case where the limitador CR is managed by another custom controller (due to different owner references), I think a log or a even better a status on the kuadrant CR noting this would be nice, since it is not obvious to the user that the limitador spec defined in the Kuadrant CR will not be reconciled due to this

**Review by KevFan — COMMENTED** [review]
Looks good generally 👍  Have 2 comments 🧑‍💻

**Review by KevFan — APPROVED** [review]
Verified and code looks good to me as we've decided to have a follow up ticket instead on if the edge case of the limitador used is owned by another custom controller 👍 

There's failing integration tests, but not sure it related to this PR 🤷

**Comment by Boomatang** [comment]
The failing integration tests where cause by the edge case that was found during this time. Where a second kuadrant CR was trying to take ownership of the limitador CR, which was owned by the first kuadrant CR. In this case the integration tests required a small refactor.

File: controllers/kuadrant_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (didierofrivia) [reply controllers/kuadrant_controller.go]: I agree with Kevin here, and to add a bit more, it could be a common scenario to have a _limitador_ instance in the cluster with its default name `limitador`, thus Kuadrant should not ownership if not explicitly declared in the Kuadrant CR.

**Comment by didierofrivia** [comment]
I'm getting the following error when running `make local-setup`

```bash
error: no objects passed to apply
make[4]: *** [sail-install] Error 1
make[3]: *** [istio-install] Error 2
make[2]: *** [local-cluster-setup] Error 2
make[1]: *** [local-env-setup] Error 2
make: *** [local-setup] Error 2
```

**Comment by Boomatang** [comment]
@didierofrivia The istio-operator removed the manifests on Friday.  Its an error on related to these changes. You can try `ISTIO_INSTALL_SAIL=false make local-setup`, it worked for me on a different PR this morning.

https://github.com/maistra/istio-operator/pull/1576

**Comment by adam-cattermole** [comment]
> I'm getting the following error when running `make local-setup`
> 
> ```shell
> error: no objects passed to apply
> make[4]: *** [sail-install] Error 1
> make[3]: *** [istio-install] Error 2
> make[2]: *** [local-cluster-setup] Error 2
> make[1]: *** [local-env-setup] Error 2
> make: *** [local-setup] Error 2
> ```

This error is from sail (again) #400, addressed partially in #395. For now you can run with `ISTIO_INSTALL_SAIL=false`

File: controllers/kuadrant_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Boomatang) [reply controllers/kuadrant_controller.go]: I agree it could be common to have any existing limitador CRs called `limitador` on the cluster. But let me ask this question. If you have a limitador instance that you do not want kuadrant to manage, why would you create the kuadrant CR in the same namespace as the existing limitador CR? Next question I would ask, can limitador work with two limitador CRs in the same namespace?

**Review by didierofrivia — APPROVED** [review]
works as expected! noted regarding the edge case, and probably we might want to address limitador instance naming too in that following issue (?) 
Good job! 🎖️
